7879: Automated Analysis taking very long time to produce results for Class Leak Rule and showing wrong results.

Suchitainf opened the pull request:
This PR takes care of optimizing and correcting the Class leak rule results on Automated Analysis Page.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JMC-7879](https://bugs.openjdk.org/browse/JMC-7879): Automated Analysis taking very long time to produce results for Class Leak Rule and showing wrong results.


### Reviewers
 * [Henrik Dafgård](https://openjdk.org/census#hdafgard) (@Gunde - **Reviewer**) ⚠️ Review applies to [56a90a79](https://git.openjdk.org/jmc/pull/419/files/56a90a79d2bb8fea72a140c7f1296a18c9af15dc)
 * [Brice Dutheil](https://openjdk.org/census#bdutheil) (@bric3 - Author) ⚠️ Review applies to [292b2251](https://git.openjdk.org/jmc/pull/419/files/292b2251c027dc9fc8e12ea017c571b4d8270dd2)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmc pull/419/head:pull/419` \
`$ git checkout pull/419`

Update a local copy of the PR: \
`$ git checkout pull/419` \
`$ git pull https://git.openjdk.org/jmc pull/419/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 419`

View PR using the GUI difftool: \
`$ git pr show -t 419`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmc/pull/419.diff">https://git.openjdk.org/jmc/pull/419.diff</a>

</details>

aptmac reviewed (commented):
Is https://bugs.openjdk.org/browse/JMC-7879 a duplicate of https://bugs.openjdk.org/browse/JMC-7248?

It looks like this PR is https://github.com/openjdk/jmc/pull/248 with some of the review comments addressed, should the former PR be closed in favor of this?

There are a couple of review comments that didn't receive any changes in this PR:
https://github.com/openjdk/jmc/pull/248/files#r756349943
https://github.com/openjdk/jmc/pull/248/files#r756353644

It looks like the consensus on https://github.com/openjdk/jmc/pull/248 was that it was good enough to go in, maybe ping @bric3 and/or @Gunde for a re-review/approval?

Suchitainf commented:
> Is https://bugs.openjdk.org/browse/JMC-7879 a duplicate of https://bugs.openjdk.org/browse/JMC-7248?
> 
> It looks like this PR is #248 with some of the review comments addressed, should the former PR be closed in favor of this?
> 
> There are a couple of review comments that didn't receive any changes in this PR: https://github.com/openjdk/jmc/pull/248/files#r756349943 https://github.com/openjdk/jmc/pull/248/files#r756353644
> 
> It looks like the consensus on #248 was that it was good enough to go in, maybe ping @bric3 and/or @Gunde for a re-review/approval?

This PR is a superset of the previous one. The older one deals with only optimization of the class leak rule but this one deals with the optimization as well as the correction of the results shown on automated analysis page for class leak rule. I have implemented all the review comments given on the previous PR. With respect to the two comments mentioned above, the configured timeout is being fetched from preference hence I am finding it difficult to make it final. The interrupted exception handling is done where the method is called in catch block. I had already pinged about this issue in general slack channel and it was discussed in hangout also. 

@Gunde and @bric3 Request both of you to please take a look at this PR and let me know your review comments. Thanks in advance.

bric3 commented:
@Suchitainf Sorry for the lack of response, I was on vacation in August, and got some stuff at work to handle, I should have more luck next week.

Gunde commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/general/ClassLeakingRule.java:
```diff
@@ -147,7 +153,7 @@ public Map<String, EventAvailability> getRequiredEvents() {
 	private IResult getResult(
 		IItemCollection items, IPreferenceValueProvider valueProvider, IResultValueProvider dependencyResults) {
 		int warningLimit = (int) valueProvider.getPreferenceValue(WARNING_LIMIT).longValue();
-
+		CONFIGURED_TIMEOUT = (int) valueProvider.getPreferenceValue(MAX_TIMEOUT).longValue();
```
Nit: This preference should use .clampedLongValueIn and passing in the appropriate unit, so since the timeout use of this specifies TimeUnit.MINUTES we should use valueProvider.getPreferenceValue(MAX_TIMEOUT).clampedLongValueIn(UnitLookup.MINUTE).longValue().

Suchitainf replied:
If we use clampedLongValueIn then longValue() should be removed right? We can't use both together.

Gunde replied:
Yes, sorry about that confusion.

Gunde commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java:
```diff
@@ -96,20 +105,36 @@ private void calculateData(IItemCollection input) {
-					}
-					data.add(row);
+				} finally {
+					exec.shutdown();
+					exec.awaitTermination(ClassLeakingRule.CONFIGURED_TIMEOUT, TimeUnit.MINUTES);
 				}
```
This timeout should be passed as a parameter instead as this could differ if rules are run concurrently with different configurations for this. Also if we would want to use this method from another rule it would be impossible to set the timeout properly.

Gunde reviewed (changes requested):
I have a few nits and one necessary change before we can merge this.

Gunde commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java:
```diff
@@ -96,20 +105,36 @@ private void calculateData(IItemCollection input) {
-						// Optimization - it is too expensive to do aggregation for these. You simply
-						// get first non-null matching attribute - we're only using this for the group by today.
-						row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+				ExecutorService exec = Executors.newWorkStealingPool();
```
We might want to set a target level parallelism here?

bric3 commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java:
```diff
@@ -96,20 +105,36 @@ private void calculateData(IItemCollection input) {
+								for (int j = 0; j < aggregators.size(); j++) {
+									row[column + j] = rowCollection.getAggregate(aggregators.get(j));
+								}
+								synchronized (data) {
+									data.add(row);
+								}
```
Last time I didn't thought about the order of the results, Is it something to take care, I really lack flightrecorder rule knowledge here.

bric3 commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java:
```diff
@@ -96,20 +105,36 @@ private void calculateData(IItemCollection input) {
-						row[column + j] = rowCollection.getAggregate(aggregators.get(j));
-					}
-					data.add(row);
+				} finally {
+					exec.shutdown();
+					exec.awaitTermination(ClassLeakingRule.CONFIGURED_TIMEOUT, TimeUnit.MINUTES);
```
I believe the proper executor shutdown sequence would be something like this

```suggestion
				} finally {
				    exec.shutdown();
					try {
						if (!exec.awaitTermination(60, TimeUnit.SECONDS)) {
							exec.shutdownNow();
							
							// possibly wait for other task to respond to cancellation, possibly not needed in this case. But useful to keep, if this code gets extracted away.
							if (!pool.awaitTermination(timeout.clampedLongValueIn(MINUTE), TimeUnit.MINUTES)) {
								// TODO notify error that pool did not terminate
							}
						}
					} catch (InterruptedException ie) {
						exec.shutdownNow();
						Thread.currentThread().interrupt();
					}
```

Gunde commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/general/ClassLeakingRule.java:
```diff
@@ -101,8 +101,13 @@ public class ClassLeakingRule implements IRule {
 			Messages.getString(Messages.General_CONFIG_CLASS_LIMIT_LONG), NUMBER, NUMBER_UNITY.quantity(5));
 
+	public static final TypedPreference<IQuantity> MAX_TIMEOUT = new TypedPreference<>(
+			"classLeaking.calculation.timeout", //$NON-NLS-1$
```
`classLeakingRule`, no?

Suchitainf replied:
I have checked at other places and "Rule" is not mentioned anywhere for Preferences. For ClassLeakRule also there are other preferences like classLeaking.warning.limit and classLeaking.classesToReport.limit , the keyword Rule is not used anywhere. Hence, I have skipped it to follow the convention. For other rules also like Biased Revocation Rule, the preference is biasedRevocation.classesToReport.limit.

bric3 replied:
> `classLeakingRule`, no?

Good catch on the typo

bric3 replied:
@Suchitainf I missed those from GH, you can dismiss this suggestion !

Suchitainf commented:
@Gunde @bric3 I have tried to implement all the review comments. Please re-review the PR and let me know if there is something which is still need to be corrected.

bric3 commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/ItemResultSetFactory.java:
```diff
@@ -40,6 +40,10 @@
 public class ItemResultSetFactory {
 	public IItemResultSet createResultSet(IItemCollection items, IItemQuery query) {
-		return new DefaultIItemResultSet(items, query);
+		return new DefaultIItemResultSet(items, query, 0);
```
The issue with a value of 0 means the termination will be effective immediately, I think you can choose either
* a guard value like `-1` is better and in `DefaultIItemResultSet` just skip the termination.
* chose a very big value like one day (that's what JDK 19 does in `ExecutorService.close`, but the code can even wait longer)

bric3 commented:
@Suchitainf There's still these two comment to address in my opinion: 
* https://github.com/openjdk/jmc/pull/419#discussion_r987769827
* https://github.com/openjdk/jmc/pull/419#discussion_r988313757

Suchitainf commented:
@bric3 We need to show the names of top 5 classes which were loaded maximum times without being unloaded. So the ordering matters here. That's why I have skipped that comment.

Gunde commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/general/ClassLeakingRule.java:
```diff
@@ -101,8 +101,13 @@ public class ClassLeakingRule implements IRule {
+			"classLeaking.calculation.timeout", //$NON-NLS-1$
+			Messages.getString(Messages.ClassLeakingRule_CONFIG_CALCULATION_TIMEOUT),
+			Messages.getString(Messages.ClassLeakingRule_CONFIG_CALCULATION_TIMEOUT_LONG), NUMBER,
+			NUMBER_UNITY.quantity(5));
```
🤦 
Of course that's right, a timespan is the correct type here.

bric3 commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java:
```diff
@@ -134,7 +134,8 @@ public void run() {
 							exec.shutdownNow();
-							exec.awaitTermination(configuredTimeout, TimeUnit.MINUTES);
+							if (configuredTimeout != 0)
+								exec.awaitTermination(configuredTimeout, TimeUnit.MINUTES);
```
Sorry looking at this now I think the proposition (1) in [comment](https://github.com/openjdk/jmc/pull/419#discussion_r988313757) is incorrect

* ~~a guard value like `-1` is better and in `DefaultIItemResultSet` just skip the termination.~~ 

If the value is `0` (or the `-1` which conveys to human this is an invalid value and a such disable the timeout) the code should just wait that all results have been delivered. Also maybe make this guard value a constant that can be reused in `ItemResultSetFactory`.

Instead the other proposition, to have a very large termination timeout, might be more appropriate in the present case, JDK 19's `ExecutorService.close()` for example use a timeout of _1 day_.

bric3 reviewed (commented):
> @bric3 We need to show the names of top 5 classes which were loaded maximum times without being unloaded. So the ordering matters here. That's why I have skipped that comment.

I'm a bit confused, as in `DefaultIItemResultSet` the _insertion_ in `data` order is dependent on the iteration order of the aggregate collection, since it's a set, there's no ordering guarantees. Although I lack knowledge of the actual type and the implementation is a `TreeSet` or something alike.

https://github.com/openjdk/jmc/blob/1696e03a1fddb9cc68d32bf21a281ef24366e552/core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java#L95-L99

Moreover the usage of the thread pool here will not guarantee the `add` operation will happen in the same order as the tasks have been submitted. Which leads me to think that the code should be reworked here to ensure that the results will be inserted in order.

Suchitainf commented:
> > @bric3 We need to show the names of top 5 classes which were loaded maximum times without being unloaded. So the ordering matters here. That's why I have skipped that comment.
> 
> I'm a bit confused, as in `DefaultIItemResultSet` the _insertion_ in `data` order is dependent on the iteration order of the aggregate collection, since it's a set, there's no ordering guarantees. Although I lack knowledge of the actual type and the implementation is a `TreeSet` or something alike.
> 
> https://github.com/openjdk/jmc/blob/1696e03a1fddb9cc68d32bf21a281ef24366e552/core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java#L95-L99
> 
> Moreover the usage of the thread pool here will not guarantee the `add` operation will happen in the same order as the tasks have been submitted. Which leads me to think that the code should be reworked here to ensure that the results will be inserted in order.

I have used a synchronized block for data. After that I am getting consistent results. Without that I was getting different results every time. I mean when I was working for test JFR wldf.jfr without synchronized block on data, the top 5 results varied every time.

bric3 commented:
> I have used a synchronized block for data. After that I am getting consistent results. Without that I was getting different results every time. I mean when I was working for test JFR wldf.jfr without synchronized block on data, the top 5 results varied every time.

Have you tried with an intermediate `ConcurrentLinkedQueue`, then copy the results in the data array list ?

Suchitainf commented:
@bric3 I have tried two intermediate data structures but both implementations are giving wrong results.

Approach 1:
private final List<Object[]> processingList = Collections.synchronizedList(new ArrayList<Object[]>()); //Creation of intermediate data structure List
processingList.add(row); //Adding row without synchronized block
OR
synchronized (processingList) { //Adding row with synchronized block
processingList.add(row);
}
data.addAll(processingList); // Finally adding the processed list to the Arraylist.


Approach 2:
private final ConcurrentLinkedQueue<Object[]> processingQueue = new ConcurrentLinkedQueue<Object[]>(); //Creation of intermediate data structure ConcurrentLinkedQueue
processingQueue (row); //Adding row without synchronized block
OR
synchronized processingQueue { //Adding row with synchronized block
processingQueue (row);
}
data.addAll(processingQueue); // Finally adding the processed Queue to the Arraylist.

To be more precise, by wrong results, I mean the size of data structure (I am trying printing while debugging) and the final top 5 results are not consistent.

Suchitainf commented:
This PR is with respect to an issue raised by one of our customers where Class Leak Rule is taking 60-90 minutes to evaluate the results. The issue was raised on Sat Apr 24 2021. Its been more than a year and we couldn't give any solution to our customers, they are still living with that unacceptable calculation time. Now they are chasing us with several messages and mails to understand when will they receive a fix for it. They have marked it as P2.

Hence, considering below points:
1. We are reducing the rule calculation time from 60-90 mins to less than 60 seconds. These calculation timings are with respect to the JFR shared by customer while raising the issue in BugDB (Oracle Internal portal for raising issues).
2. We are correcting the results shown for Class Leak Rule and the correction is a drastic change. The existing approach is showing everything in GREEN even if there is a big issue and HIGH ALERT should be given as part of this rule results. This PR is fixing that issue and showing the correct result to the customers.
3. We have a fix which is not breaking anything as such. No Critical or Must DO review comment is left out unattended.

Is it possible to accept this PR for 8.3 release and create a new lesser priority enhancement request if there are any modifications required with respect to re-designing of this Rule or using some other data structure?

@thegreystone @Gunde @aptmac @bric3 I would like to know your views on this. Thanks in advance.

thegreystone commented:
@aptmac, @Gunde & @bric3 - can you please help Suchita complete this PR?

aptmac commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java:
```diff
@@ -59,19 +62,26 @@ final class DefaultIItemResultSet implements IItemResultSet {
-	private void calculateData(IItemCollection input) {
+	private void calculateData(IItemCollection input, int configuredTimeout) throws InterruptedException {
 		input = input.apply(query.getFilter());
+		final IItemCollection newInput = input;
```
Very minor nit, but this could be moved to be inside the else block where it's used? It's not required for the first chunk of the if-statement.

aptmac reviewed (commented):
I think it looks okay for me. It would be nice to have a way of reproducing the issue to verify, but I get it's tough if the original jfr is from a customer and can't be shared.

Having said that, @Gunde and @bric3 did substantial review on this PR as well as the previous (https://github.com/openjdk/jmc/pull/248), and while on the surface it looks like everything has been addressed I would feel much more comfortable if they could just confirm that what they wanted addressed was updated in this PR before merging it.

bric3 reviewed (commented):
I'm still not at ease there. I think contributing an actual test case for this rule (and/or the `DefaultIItemResultSet`) might beneficial.

Having a synchronized block wrapping the `add(row)`, looks suspicious as this somehow coordinate threads but this behavior still non-determinist as it depends on how threads get scheduled. 

As per this [comment](https://github.com/openjdk/jmc/pull/419#issuecomment-1273226601)
> To be more precise, by wrong results, I mean the size of data structure (I am trying printing while debugging) and the final top 5 results are not consistent.

I do believe this happen due to the reason mentioned above. Maybe the resulting list needs to be sorted then ?

Suchitainf commented:
@aptmac @bric3 I have implemented the review comments suggested. 

@bric3 Yes, you are right " In DefaultIItemResultSet the insertion in data order is dependent on the iteration order of the aggregate collection, since it's a set, there's no ordering guarantees. " We are just processing the records in DefaultIItemResultSet. Sorting is taken care in ClassLeakingRule.java class at line no 182 ( entries.sort(null); ) after all the records are processed. 

The reason for why wrong records were shown previously was : We were showing the results in ascending order of the count i.e. difference between the number of times classed loaded vs unloaded. Whereas we had to show in descending order. For Ex: There are 10 classes A (0), B (1), C (1)........H (55), I (60), J (75). As per current implementation we are showing A, B, C on class leak rule results on Automated Analysis Screen and saying everything is normal. Whereas, we have to show  J, I, H classes because they are loaded without being unloaded and they need attention. So JMC has to show warning instead of showing OK results. 

Regarding test case, we already have a JFRRuleBaseline.xml for testing rules and their results and I have updated that as part of this PR.

Suchitainf commented:
Hi @bric3 @Gunde @aptmac , Did you get a chance to check the latest patch? Please let me know what else is required, so that we can close this PR.

Suchitainf commented:
In the latest patch, I have removed the synchronized block which was the main bottle neck for the approval of this PR. Also, used the suggested data structure i.e. ConcurrentLinkedQueue for the processing of records. 

I have debugged in detail and figured out that the reason for inconsistent results for the customer JFR which has around 65000 class entries to process was the time multithreaded tasks were taking to complete which was dependent on many factors like how many cores our CPU has OR how busy our CPU is when its performing those tasks. Hence, for safer side I have added a check to see whether all tasks are performed or not. If not, I have added a console message to alert the user for increasing the configured timeout. The default timeout is 5 min which is quite sufficient to complete all the tasks still if the count of entries are too much and there are some pending tasks, user will be alerted for the same.

bric3 commented:
Hi @Suchitainf I'll have a look tommorrow.

bric3 commented on core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java:
```diff
@@ -89,29 +102,69 @@ private void calculateData(IItemCollection input) {
+		}
+
+		if (completedTask < totalAssignedTasks.size()) {
+			System.out.println(
+					"The results may be inaccurate as JMC is unable to process all the class entries to determine "
+							+ "the class leak results. Please increase the configured timeout in preferences to see the accurate results.  ");
```
I believe this should be changed to use the jul logger. Other flighrecorder component uses that.

E.g.

```java
private final static Logger LOGGER = Logger.getLogger("org.openjdk.jmc.flighrecorder.rules.jdk.util");
```

Suchitainf replied:
Hi @bric3 Thanks for taking out time and reviewing the PR. I have replaced sysout with Logger now.

bric3 reviewed (approved):
OK for me. There's one tidbit I think should be changed is using the jul logger over `System.out`.

Suchitainf commented:
/integrate
